### PR TITLE
CDAP-13154 Use just the username and not kerberos principal for auth…

### PR DESCRIPTION
**Pull request by rohitsinha54**
…orization

For impersonation a kerberos principal can be provided in multiple forms. For example during namespace creation a user might choose to provide the kerberos principal which matches the environment and keytabs. For more details on different forms of kerbero principal refer: https://docs.oracle.com/cd/E21455_01/common/tutorials/kerberos_principal.html

Now during authorization when there is an impersonation we do authorization on the impersonated user principal. At this point it is important we differentiate between different forms of kerberos principal and do the authorization just on the username (primary part of kerberos principal) and not on the complete principal string

Issue: https://issues.cask.co/browse/CDAP-13154
Build: https://builds.cask.co/browse/CDAP-RUT1499-1

### CODE REVIEW

File: cdap-security/src/main/java/co/cask/cdap/security/authorization/AuthorizationUtil.java [hunk 1]
```diff
@@ -52,12 +53,15 @@
  */
 public class AuthorizationUtil {
```
Comment (chtyim) [inline cdap-security/src/main/java/co/cask/cdap/security/authorization/AuthorizationUtil.java]: Util class should be `final`.

**Review by chtyim — COMMENTED** [review]
Just one minor comment, rest LGTM.

File: cdap-security/src/main/java/co/cask/cdap/security/authorization/AuthorizationUtil.java [hunk 2]
```diff
@@ -160,7 +164,17 @@ public static String getAppAuthorizingUser(OwnerAdmin ownerAdmin, Authentication
...
-    return effectiveOwner != null ? effectiveOwner.getPrincipal() : authenticationContext.getPrincipal().getName();
+
+    // CDAP-13154 If impersonation is configured for either the application or namespace the effective owner will be
+    // a kerberos principal which can have different form
+    // (refer: https://docs.oracle.com/cd/E21455_01/common/tutorials/kerberos_principal.html). For example it can be
+    // a complete principal name (alice/somehost.net@someREALM). For authorization we need the enforcement to happen
+    // on the username and not the complete principal. The user name is the shortname of the principal so return the
+    // shortname as authorizing user.
+    String appAuthorizingUser = effectiveOwner != null ?
+      new KerberosName(effectiveOwner.getPrincipal()).getShortName() : authenticationContext.getPrincipal().getName();
```
Comment (anew) [inline cdap-security/src/main/java/co/cask/cdap/security/authorization/AuthorizationUtil.java:175]: don't we need to use getShortName() also for the else case?

File: cdap-security/src/test/java/co/cask/cdap/security/authorization/AuthorizationUtilTest.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    // Note: it is important to initialize the UGI before we call AuthorizationUtil.getAppAuthorizingUser(..)
+    // which uses KerberosName since KerberosName expect the rules for matching to be set. See
+    // http://lucene.472066.n3.nabble.com/KerberosName-rules-are-null-during-KerberosName-getShortName-
+    // in-KerberosAuthenticationHandler-td4074309.html for more context
+    username = UserGroupInformation.getCurrentUser().getShortUserName();
+    namespaceClient = new InMemoryNamespaceClient();
+    authenticationContext = new AuthenticationTestContext();
+  }
+
+  @org.junit.Test
```
Comment (anew) [inline cdap-security/src/test/java/co/cask/cdap/security/authorization/AuthorizationUtilTest.java]: import this and replace with @Test

File: cdap-security/src/test/java/co/cask/cdap/security/authorization/AuthorizationUtilTest.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    principal = username + "@REALM.net";
+    nsMeta = new NamespaceMeta.Builder().setName(namespaceId).setPrincipal(principal)
+      .setKeytabURI("doesnotmatter").build();
+    namespaceClient.create(nsMeta);
+    Assert.assertEquals(username, AuthorizationUtil.getAppAuthorizingUser(ownerAdmin, authenticationContext,
+                                                                          applicationId, null));
+    // clean up
+    namespaceClient.delete(namespaceId);
+  }
+
```
Comment (anew) [inline cdap-security/src/test/java/co/cask/cdap/security/authorization/AuthorizationUtilTest.java:89]: you should also test the case where the authentication context has a principal of each of the three forms (and the namespace has no principal). AuthenticationTestContext only provides short username as principal.

**Review by anew — COMMENTED** [review]
please see comments

File: cdap-security/src/main/java/co/cask/cdap/security/authorization/AuthorizationUtil.java (hunk 2 above)
Comment (rohitsinha54) [reply cdap-security/src/main/java/co/cask/cdap/security/authorization/AuthorizationUtil.java:175]: the principal is always shortname. The AuthenticationContext is populated by the sign-in where we support just username and not principal to sign in.

File: cdap-security/src/test/java/co/cask/cdap/security/authorization/AuthorizationUtilTest.java (hunk 3 above)
Comment (rohitsinha54) [reply cdap-security/src/test/java/co/cask/cdap/security/authorization/AuthorizationUtilTest.java]: oops!

File: cdap-security/src/test/java/co/cask/cdap/security/authorization/AuthorizationUtilTest.java (hunk 4 above)
Comment (rohitsinha54) [reply cdap-security/src/test/java/co/cask/cdap/security/authorization/AuthorizationUtilTest.java:89]: I don't think AuthenticationContext can be set by kerberos principal. Its set with the logged in username from our authentication module. We don't support logging in with kerberos principal as of now.

**Review by anew — APPROVED** [review]

**Pull request merged**